### ISSUE TITLE
ClassNotFoundException While ShuttingDown Apache Felix Using H2

### ISSUE THREAD

**Issue by joebrassi** [brettwooldridge/HikariCP#298]
So far I have tested HirakiCP with Hsqldb, H2 and MySql. It works on all three of them, but with H2, HirakiCP throws the following exception when the Apache Felix is shutting down: 

```console
g! ^D
gosh: stopping framework
Exception in thread "Thread-0" java.lang.NoClassDefFoundError: org/h2/mvstore/MVMap$2
        at org.h2.mvstore.MVMap.entrySet(MVMap.java:865)
        at org.h2.store.LobStorageMap.removeAllForTable(LobStorageMap.java:294)
        at org.h2.engine.Database.removeOrphanedLobs(Database.java:1302)
        at org.h2.engine.Database.close(Database.java:1223)
        at org.h2.engine.DatabaseCloser.run(DatabaseCloser.java:63)
Caused by: java.lang.ClassNotFoundException: Unable to load class 'org.h2.mvstore.MVMap$2' because the bundle wiring for org.h2 is no longer vali
d.
        at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1523)
        at org.apache.felix.framework.BundleWiringImpl.access$400(BundleWiringImpl.java:79)
        at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:1998)
        at java.lang.ClassLoader.loadClass(Unknown Source)
        ... 5 more
```

This seems like a race condition between the HirakiCP bundle and the H2 bundle, where HirakiCP is making a call to an already-stopped bundle (H2), hence throwing a `NoClassDefFoundError` exception. Replacing the persistence provider by Hsqldb or Mysql works just fine, no exception.

Versions used:
* `h2-1.4.186.jar`
* `HikariCP-2.3.5.jar`

**Comment by brettwooldridge** [comment]
Is that the full stack?  Where is HikariCP involved in this?

**Comment by joebrassi** [comment]
Yes, that is the full stack. The exception is thrown because HirakiCP is trying to finalize the pool created but with H2 it fails to finish it. I have tested several combinations which include Apache-DBCP, HirakiCP, H2, Hsqldb and MySql, and the only combination that throws exception is HirakiCP and H2.

**Comment by joebrassi** [comment]
The previous stack trace is from Apache Felix console. Running some integration tests I get this stack trace:

```console
org.h2.message.DbException: Log file error: "D:/projects/cells/sources/org-sample3/target/osgi-runner/data/h2_hikari_dba.trace.db", cause: "java.lang.NullPointerException" [90034-186]
	at org.h2.message.DbException.get(DbException.java:168)
	at org.h2.message.TraceSystem.logWritingError(TraceSystem.java:276)
	at org.h2.message.TraceSystem.openWriter(TraceSystem.java:297)
	at org.h2.message.TraceSystem.writeFile(TraceSystem.java:245)
	at org.h2.message.TraceSystem.write(TraceSystem.java:229)
	at org.h2.message.Trace.error(Trace.java:151)
	at org.h2.engine.DatabaseCloser.run(DatabaseCloser.java:69)
Caused by: org.h2.jdbc.JdbcSQLException: Log file error: "D:/projects/wiredcells/sources/com.wiredcells.persistence/data/target/osgi-runner/data/h2_hikari_dba.trace.db", cause: "java.lang.NullPointerException" [90034-186]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:345)
	... 7 more
Caused by: java.lang.NullPointerException
	at org.apache.felix.framework.BundleWiringImpl.diagnoseClassLoadError(BundleWiringImpl.java:2625)
	at org.apache.felix.framework.BundleWiringImpl.access$700(BundleWiringImpl.java:75)
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:1967)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at org.h2.util.IOUtils.getBufferedWriter(IOUtils.java:407)
	at org.h2.message.TraceSystem.openWriter(TraceSystem.java:293)
	... 4 more
org.h2.message.DbException: Log file error: "D:/projects/wiredcells/sources/com.wiredcells.persistence/data/target/osgi-runner/data/h2_hikari_dba.trace.db", cause: "java.lang.NullPointerException" [90034-186]

```

**Comment by brettwooldridge** [comment]
This is basically a start-level issue.  See here:

http://eclipsesource.com/blogs/2009/06/10/osgi-and-start-levels/

and here...
```
When the Framework is shutdown the Framework's start-level is decremented to zero.
For each start-level the framework implementation can select an appropriate ordering
for stopping the bundles in the start-level. This may include stopping bundles in reverse
dependency order such that all bundles which depend on a given bundle are shutdown
before the given bundle. In case of dependency cycles, the bundles shutdown order
would be unspecified. This helps prevent class load requests from lazy activation bundles
that have been stopped during the shutdown process.
```
http://www.osgi.org/Design/LazyStart

The fact that DBCP may or may not encounter this issue is immaterial (maybe they don't shutdown properly, who knows).  The application (yours) defines the start order, and bundles are stopped in reverse order.  HikariCP needs to start after H2, and thus will stop before it.

**Comment by joebrassi** [comment]
Indeed, that may be a solution but why then it does not happen with Hsqldb or Mysql? There is no race condition with these providers.

**Comment by brettwooldridge** [comment]
I'm not sure of the causes of either one of those errors.  It is quite possible that it is not happening with Hsqldb or Mysql because all of the classes they reference during shutdown have already been loaded.  In the case of H2, at least when HikariCP is driving it, it seems that the class ``org.h2.mvstore.MVMap$2`` is first referenced after the bundle has already uninstalled.  Could be a timing issue, or something else -- debugging OSGi classloading is no fun.  And you might not have to if you control the start-order.  Without controlling the start/stop order, it seems you are taking a gamble.  Even if it worked one hundred times in a row with no error, it could fail on the 101st.

**Comment by joebrassi** [comment]
I was going to dig into the HikariCP code to try to find out what was causing it, but I already have plenty of work to also get into these details. I already did it with Apache-DBCP-2.1, it does not work out-the-box in OSGi, I had to fix the code in a couple of places in order to make it work, it simply won't even let you get a connection as it is. In the case of HikariCP, it is a harmless exception, happens only at Felix shutdown, but even in that case it gives that bad taste to the final user of having a buggy code. I would change the start order but in my case that is more a workaround than a solution. Anyways, I let you know if I find the issue causing this exception.

**Comment by brettwooldridge** [comment]
Controlling OSGi start/stop order is not a workaround, it is a necessity.  OSGi can infer start order for bundles that declare dependencies to packages exported by other bundles, but when two bundles (H2 and HikariCP) have no explicit relationship, and instead have an implicit relationship, it is up to the application to configure OSGi's start order.

In this case, because OSGi does not know that HikariCP has a runtime dependency on H2, the framework treats them as having equal start-level (default 6) in which case the specification says that shutdown order is not guaranteed.  Felix is shutting down H2 before HikariCP, which is the source of the exception.

The only way to avoid this, and the correct way, is to specify that HikariCP has a higher start-level than H2.  This will ensure that Felix starts HikariCP after H2, and stops HikariCP before stopping H2.

**Comment by joebrassi** [comment]
Ok, found the issue. Seems like it was not a race condition, I tried changing the start levels of both bundles `h2` and `hikariCP` but the exception was still there. Here is the execution:

```console
D:\projects\cells\runner\felix>java -jar bin/felix.jar
[FelixStartLevel] INFO com.zaxxer.hikari.HikariDataSource - HikariCP pool HikariPool-0 is starting.
____________________________
Welcome to Apache Felix Gogo

g! lb
START LEVEL 1
   ID|State      |Level|Name
    0|Active     |    0|System Bundle (4.6.0)
    1|Active     |    1|cells.persistence (0.1.0)
    2|Active     |    1|Apache Commons DBCP (2.1.0)
    3|Active     |    1|Apache Commons Logging (1.2.0)
    4|Active     |    1|Apache Commons Pool (2.3.0)
    5|Active     |    1|HSQLDB (2.3.2)
    6|Active     |    1|Javassist (3.19.0.GA)
    7|Active     |    1|Apache Log4j (1.2.17)
    8|Active     |    1|Oracle Corporation' JDBC Driver for MySQL (5.1.27)
    9|Active     |    1|Apache Felix Bundle Repository (2.0.2)
   10|Active     |    1|Apache Felix Gogo Command (0.14.0)
   11|Active     |    1|Apache Felix Gogo Runtime (0.12.1)
   12|Active     |    1|Apache Felix Gogo Shell (0.10.0)
   13|Active     |    1|osgi.cmpn (5.0.0.201305092017)
   14|Active     |    1|slf4j-api (1.7.10)
   15|Resolved   |    1|slf4j-simple (1.7.10)
   16|Resolved   |    2|H2 Database Engine (1.4.186)
   17|Resolved   |    3|HikariCP (2.3.5)
g! ^D
gosh: stopping framework
Exception in thread "Thread-0" java.lang.NoClassDefFoundError: org/h2/mvstore/MVMap$2
        at org.h2.mvstore.MVMap.entrySet(MVMap.java:865)
        at org.h2.store.LobStorageMap.removeAllForTable(LobStorageMap.java:294)
        at org.h2.engine.Database.removeOrphanedLobs(Database.java:1302)
        at org.h2.engine.Database.close(Database.java:1223)
        at org.h2.engine.DatabaseCloser.run(DatabaseCloser.java:63)
Caused by: java.lang.ClassNotFoundException: Unable to load class 'org.h2.mvstore.MVMap$2' because the bundle wiring for org.h2 is no longer vali
d.
        at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1523)
        at org.apache.felix.framework.BundleWiringImpl.access$400(BundleWiringImpl.java:79)
        at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:1998)
        at java.lang.ClassLoader.loadClass(Unknown Source)
        ... 5 more
```

Then I added some code to *close* the data-sources opened in the `Activator.Stop(BundleContext context)` method and this solved the issue. This is the code I added and the results:

```java
public class Activator implements BundleActivator {
    HikariDataSource hikariDataSource;

    @Override
    public void start(BundleContext bundleContext) throws Exception {
    ...
    ...
    }

    @Override
    public void stop(BundleContext bundleContext) throws Exception {
        hikariDataSource.close(); // This line is mandatory when using H2
    }
}
```

And the result is:

```console
D:\projects\cells\runner\felix>java -jar bin/felix.jar
[FelixStartLevel] INFO com.zaxxer.hikari.HikariDataSource - HikariCP pool HikariPool-0 is starting.
____________________________
Welcome to Apache Felix Gogo

g! lb
START LEVEL 1
   ID|State      |Level|Name
    0|Active     |    0|System Bundle (4.6.0)
    1|Active     |    1|com.wiredcells.persistence (0.1.0)
    2|Active     |    1|Apache Commons DBCP (2.1.0)
    3|Active     |    1|Apache Commons Logging (1.2.0)
    4|Active     |    1|Apache Commons Pool (2.3.0)
    5|Active     |    1|H2 Database Engine (1.4.186)
    6|Active     |    1|HikariCP (2.3.5)
    7|Active     |    1|HSQLDB (2.3.2)
    8|Active     |    1|Javassist (3.19.0.GA)
    9|Active     |    1|Apache Log4j (1.2.17)
   10|Active     |    1|Oracle Corporation' JDBC Driver for MySQL (5.1.27)
   11|Active     |    1|Apache Felix Bundle Repository (2.0.2)
   12|Active     |    1|Apache Felix Gogo Command (0.14.0)
   13|Active     |    1|Apache Felix Gogo Runtime (0.12.1)
   14|Active     |    1|Apache Felix Gogo Shell (0.10.0)
   15|Active     |    1|osgi.cmpn (5.0.0.201305092017)
   16|Active     |    1|slf4j-api (1.7.10)
   17|Resolved   |    1|slf4j-simple (1.7.10)
g! ^D
gosh: stopping framework
[FelixStartLevel] INFO com.zaxxer.hikari.pool.HikariPool - HikariCP pool HikariPool-0 is shutting down.
```

Well, may be you would like to add this detail to your documentation, where to remove the previous exception, the bundle using HikariCP DataSource **must** close the Data Source in the `Activator.Stop()` method. This is only when HikariCP uses H2 over OSGi, with other database like hsqldb or mysql there is no need to explicitly call the `hikariDataSource.close()` method.

**Comment by brettwooldridge** [comment]
Ha!  I was assuming you were closing the HikariDataSource.  Makes more sense now.

**Comment by joebrassi** [comment]
No, I was't :) BTW, there is also a `shutdown()` method in the HirakiDataSource, does it make any difference to call one or the other (`close()` or `shutdown()`)?

**Comment by brettwooldridge** [comment]
``shutdown()`` and ``close()`` are identical.  We had ``shutdown()`` first, and then someone submitted a pull request where ``HikariDataSource`` implements the ``Closable`` interface, so ``close()`` was introduced.  For backward compatibility we didn't want to remove ``shutdown()``.

**Comment by joebrassi** [comment]
OK, perhaps for hikariCP 3.x you may remove it then.

Well, I think that's all by now. Thanks for the help. Keep you posted if I find any other detail.

**Comment by xipki** [comment]
I got the same error, even the DataSource is closed explicty. I use apache karaf 4.0.9, HikariCP 2.6.2 and H2 database 1.4.195. The maven project to reproduce the bug is attached.

Here is the error trace:
```sh
karaf@root()> install file:jars/h2-1.4.195.jar 
Bundle ID: 53
karaf@root()> install file:jars/HikariCP-2.6.2.jar 
Bundle ID: 54
karaf@root()> install file:jars/hikaricp-h2-bug-0.1.0-SNAPSHOT.jar 
Bundle ID: 55
karaf@root()> start 55
Created H2 DataSoruce
karaf@root()> list
START LEVEL 100 , List Threshold: 50
ID | State    | Lvl | Version        | Name
---------------------------------------------------------
53 | Resolved |  80 | 1.4.195        | H2 Database Engine
54 | Resolved |  80 | 2.6.2          | HikariCP
55 | Active   |  80 | 0.1.0.SNAPSHOT | hikaricp-h2-bug
karaf@root()> uninstall 
uninstall    
karaf@root()> uninstall 55
Closing H2 DataSoruce
Closed H2 DataSoruce
karaf@root()> ^Dkaraf@root()> 
Exception in thread "Thread-18" java.lang.NoClassDefFoundError: org/h2/mvstore/MVMap$2
	at org.h2.mvstore.MVMap.entrySet(MVMap.java:870)
	at org.h2.store.LobStorageMap.removeAllForTable(LobStorageMap.java:314)
	at org.h2.engine.Database.removeOrphanedLobs(Database.java:1340)
	at org.h2.engine.Database.close(Database.java:1261)
	at org.h2.engine.DatabaseCloser.run(DatabaseCloser.java:63)
Caused by: java.lang.ClassNotFoundException: Unable to load class 'org.h2.mvstore.MVMap$2' because the bundle wiring for org.h2 is no longer valid.
	at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1539)
	at org.apache.felix.framework.BundleWiringImpl.access$400(BundleWiringImpl.java:79)
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:2018)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 5 more
```
[hikaricp-h2-bug.zip](https://github.com/brettwooldridge/HikariCP/files/1046030/hikaricp-h2-bug.zip)
